File: leetcode/334. Increasing Triplet Subsequence/334.cpp
// https://helloacm.com/the-on-increasing-triplet-subsequence-algorithm/
// https://leetcode.com/problems/increasing-triplet-subsequence/

class Solution {
public:
    bool increasingTriplet(vector<int>& nums) {
        int first = INT_MAX, second = INT_MAX;
        for (const auto &n: nums) {
            if (n <= first) {
                first = n;
            } else if (n <= second) {
                second = n;
            } else return true;
        }
        return false; // can't find such triplet increasing subsequence
    }
};

File: leetcode/334. Increasing Triplet Subsequence/334.py
# https://helloacm.com/teaching-kids-programming-increasing-triplet-subsequence-algorithm/
# https://leetcode.com/problems/increasing-triplet-subsequence/
# MEDIUM, MATH

class Solution:
    def increasingTriplet(self, nums: List[int]) -> bool:
        a = b = math.inf
        for i in nums:
            if i <= a:
                a = i
            elif i <= b:
                b = i
            else:
                return True
        return False
